From 7a191d056154688c16e4997a5aa44b1588a1dea0 Mon Sep 17 00:00:00 2001
From: Sebastian Bauer <mail@sebastianbauer.info>
Date: Mon, 16 Feb 2015 22:27:07 +0000
Subject: [PATCH 3/3] Declare the __builtin_getlinearva() properly in the
 include file.

The include file is used at least twice. One pass uses it to
create an array leading to out of bounds access when this is done
outside that file.
---
 gcc/config/rs6000/rs6000-builtin.def |    8 ++++++++
 gcc/config/rs6000/rs6000.h           |    3 ---
 2 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/gcc/config/rs6000/rs6000-builtin.def b/gcc/config/rs6000/rs6000-builtin.def
index 220d1e9..d6a28d1 100644
--- a/gcc/config/rs6000/rs6000-builtin.def
+++ b/gcc/config/rs6000/rs6000-builtin.def
@@ -1963,3 +1963,11 @@ RS6000_BUILTIN_X (RS6000_BUILTIN_MTFSF, "__builtin_mtfsf",
 /* Darwin CfString builtin.  */
 BU_SPECIAL_X (RS6000_BUILTIN_CFSTRING, "__builtin_cfstring", RS6000_BTM_ALWAYS,
 	      RS6000_BTC_MISC)
+
+/* AmigaOS specific builtin. */
+RS6000_BUILTIN_2 (AMIGAOS_BUILTIN_GETLINEARVA,	/* ENUM */	\
+		    "__builtin_getlinearva",	/* NAME */	\
+		    RS6000_BTM_ALWAYS,		/* MASK */	\
+		    RS6000_BTC_MISC,		/* ATTR */	\
+		    CODE_FOR_nothing)		/* ICODE */
+
diff --git a/gcc/config/rs6000/rs6000.h b/gcc/config/rs6000/rs6000.h
index fb67b61..ad2e553 100644
--- a/gcc/config/rs6000/rs6000.h
+++ b/gcc/config/rs6000/rs6000.h
@@ -2661,9 +2661,6 @@ extern int frame_pointer_needed;
 enum rs6000_builtins
 {
 #include "rs6000-builtin.def"
-
-  /* TODO: Possibly include a subtarget builtin file */
-  AMIGAOS_BUILTIN_GETLINEARVA,
   RS6000_BUILTIN_COUNT
 };
 
-- 
1.7.10.4

